packages/react-ui/src/features/pieces/lib/steps-hooks.ts

```diff
@@ -2,63 +2,94 @@ import { useQueries, useQuery } from '@tanstack/react-query';
   Trigger,
   TriggerType,
 } from '@activepieces/shared';
 
-import { INTERNAL_ERROR_TOAST, toast } from '../../../components/ui/use-toast';
-
-import { piecesApi } from './pieces-api';
-import { getCoreActions } from './pieces-hook';
-import { stepUtils } from './step-utils';
 import {
-  PieceSelectorItem,
-  PrimitiveStepMetadata,
+  PieceGroup,
+  PieceStepMetadata,
   StepMetadata,
   StepMetadataWithStepName,
   StepMetadataWithSuggestions,
-} from './types';
-
-export const CORE_STEP_METADATA: Record<
-  Exclude<ActionType, ActionType.PIECE> | TriggerType.EMPTY,
-  PrimitiveStepMetadata
-> = {
-  [ActionType.CODE]: {
-    displayName: t('Code'),
-    logoUrl: 'https://cdn.activepieces.com/pieces/code.svg',
-    description: t('Powerful Node.js & TypeScript code with npm'),
-    type: ActionType.CODE as const,
-  },
-  [ActionType.LOOP_ON_ITEMS]: {
-    displayName: t('Loop on Items'),
-    logoUrl: 'https://cdn.activepieces.com/pieces/loop.svg',
-    description: 'Iterate over a list of items',
-    type: ActionType.LOOP_ON_ITEMS as const,
-  },
-  [ActionType.ROUTER]: {
-    displayName: 'Router',
-    logoUrl: 'https://cdn.activepieces.com/pieces/branch.svg',
-    description: t('Split your flow into branches depending on condition(s)'),
-    type: ActionType.ROUTER,
-  },
-  [TriggerType.EMPTY]: {
-    displayName: t('Empty Trigger'),
-    logoUrl: 'https://cdn.activepieces.com/pieces/empty-trigger.svg',
-    description: t('Empty Trigger'),
-    type: TriggerType.EMPTY as const,
-  },
+} from '../../../lib/types';
+
+import { piecesApi } from './pieces-api';
+import {
+  CORE_ACTIONS_METADATA,
+  CORE_STEP_METADATA,
+  stepUtils,
+} from './step-utils';
+
+const isPopularPieces = (
+  stepMetadata: StepMetadataWithSuggestions,
+  platform: PlatformWithoutSensitiveData,
+) => {
+  if (
+    stepMetadata.type !== TriggerType.PIECE &&
+    stepMetadata.type !== ActionType.PIECE
+  ) {
+    return false;
+  }
+  const popularPieces = [
+    '@activepieces/piece-gmail',
+    '@activepieces/piece-google-sheets',
+    '@activepieces/piece-openai',
+    '@activepieces/piece-schedule',
+    '@activepieces/piece-webhook',
+    '@activepieces/piece-http',
+    '@activepieces/piece-forms',
+    '@activepieces/piece-slack',
+  ];
+  const pinnedPieces = platform.pinnedPieces ?? [];
+  return [...popularPieces, ...pinnedPieces].includes(
+    (stepMetadata as PieceStepMetadata).pieceName,
+  );
+};
+
+const isFlowController = (stepMetadata: StepMetadata) => {
+  if (stepMetadata.type === ActionType.PIECE) {
+    return stepMetadata.categories.includes(PieceCategory.FLOW_CONTROL);
+  }
+  return [ActionType.LOOP_ON_ITEMS, ActionType.ROUTER].includes(
+    stepMetadata.type as ActionType,
+  );
+};
+
+const isUniversalAiPiece = (stepMetadata: StepMetadata) => {
+  if (stepMetadata.type === ActionType.PIECE) {
+    return stepMetadata.categories.includes(PieceCategory.UNIVERSAL_AI);
+  }
+  return false;
+};
+
+const getQueryKeyForStepMetadata = (
```

Comment:
I would move this below const.

People read from top to bottom, you can't tell what is this file if you start reading helper functions, I usually move them down.